v2/component/launcher.go

```diff
@@ -441,3 +511,181 @@ func (l *Launcher) RunComponent(ctx context.Context, cmd string, args ...string)
 
 	return l.metadataClient.PublishExecution(ctx, execution, outputParameters, outputArtifacts)
 }
+
+func getRuntimeArtifactSchema(rta *pipeline_spec.RuntimeArtifact) (string, error) {
+	switch t := rta.Type.Kind.(type) {
+	case *pipeline_spec.ArtifactTypeSchema_InstanceSchema:
+		return t.InstanceSchema, nil
+	case *pipeline_spec.ArtifactTypeSchema_SchemaTitle:
+		return "title: " + t.SchemaTitle, nil
+	case *pipeline_spec.ArtifactTypeSchema_SchemaUri:
+		return "", fmt.Errorf("SchemaUri is unsupported, found in RuntimeArtifact %+v", rta)
+	default:
+		return "", fmt.Errorf("unknown type %T in RuntimeArtifact %+v", t, rta)
+	}
+}
+
+func mergeRuntimeArtifacts(src, dst *pipeline_spec.RuntimeArtifact) {
+	if len(src.Uri) > 0 {
+		dst.Uri = src.Uri
+	}
+
+	if src.Metadata != nil {
+		if dst.Metadata == nil {
+			dst.Metadata = src.Metadata
+		} else {
+			for k, v := range src.Metadata.Fields {
+				dst.Metadata.Fields[k] = v
+			}
+		}
+	}
+}
+
+// TODO(neuromage): Move these helper functions to a storage package and add tests.
+func uploadFile(ctx context.Context, bucket *blob.Bucket, localFilePath, blobFilePath string) error {
+	errorF := func(err error) error {
+		return fmt.Errorf("uploadFile(): unable to complete copying %q to remote storage %q: %w", localFilePath, blobFilePath, err)
+	}
+
+	w, err := bucket.NewWriter(ctx, blobFilePath, nil)
+	if err != nil {
+		return errorF(fmt.Errorf("unable to open writer for bucket: %w", err))
+	}
+
+	r, err := os.Open(localFilePath)
+	if err != nil {
+		return errorF(fmt.Errorf("unable to open local file %q for reading: %w", localFilePath, err))
+	}
+	defer r.Close()
```

Comment:
@ajaygopinathan note, this comment is different from others, it handles an error in defer command. I am not sure when it will happen, but it seems to me that if the file fails to close, we might not have successfully written the file to disk, that's an error worth catching early.

Reply:
This is an error with closing the read-from file, so it's safe to defer. Note that for writer, we always handle the error for Close() calls.

Reply:
Ohh, you are right on this!